src/CubedSpheres/cubed_sphere_faces.jl

```diff
@@ -0,0 +1,70 @@
+using Statistics
+using Oceananigans.Architectures
+
+using OffsetArrays: OffsetArray
+
+import Base: getindex, size, show, minimum, maximum
+import Statistics: mean
+
+import Oceananigans.Fields: AbstractField, ReducedField, Field, new_data
+import Oceananigans.BoundaryConditions: FieldBoundaryConditions
+
+struct CubedSphereFaces{E, F}
+    faces :: F
+end
+
+const CubedSphereData = CubedSphereFaces{<:OffsetArray}
+
+const CubedSphereField = AbstractField{X, Y, Z, <:Union{Nothing, AbstractArchitecture}, <:ConformalCubedSphereGrid} where {X, Y, Z}
+const CubedSphereFaceField = AbstractField{X, Y, Z, A, <:ConformalCubedSphereFaceGrid} where {X, Y, Z, A}
+const CubedSphereReducedField = ReducedField{X, Y, Z, A, D, <:ConformalCubedSphereFaceGrid} where {X, Y, Z, A, D}
+
+# There must be a way to dispatch in one function without ambiguity with `new_data.jl`...
+
+function new_data(FT, arch::AbstractCPUArchitecture, grid::ConformalCubedSphereGrid, (X, Y, Z))
+    faces = Tuple(new_data(FT, arch, face_grid, (X, Y, Z)) for face_grid in grid.faces)
+    return CubedSphereFaces{typeof(faces[1]), typeof(faces)}(faces)
+end
+
+function new_data(FT, arch::AbstractGPUArchitecture, grid::ConformalCubedSphereGrid, (X, Y, Z))
+    faces = Tuple(new_data(FT, arch, face_grid, (X, Y, Z)) for face_grid in grid.faces)
+    return CubedSphereFaces{typeof(faces[1]), typeof(faces)}(faces)
+end
```

Comment:
I guess if you want to share code here you have to define a new function like `_new_data` and resolve to that for both cases. Not necessary...